ngramoffset.go

```diff
@@ -0,0 +1,346 @@
+				// and are duplicated in the normal unicode entries.
+				ngramsUnicode = append(ngramsUnicode, ng, ng)
+				offsetsUnicode = append(offsetsUnicode, offsets[i], offsets[i+1])
+			}
+		} else {
+			if i == len(ngrams)-1 || ngrams[i+1]&ngramAsciiMask == ngrams[i+1] {
+				ngramsUnicode = append(ngramsUnicode, ng, ng)
+				offsetsUnicode = append(offsetsUnicode, offsets[i], offsets[i+1])
+			} else {
+				ngramsUnicode = append(ngramsUnicode, ng)
+				offsetsUnicode = append(offsetsUnicode, offsets[i])
+			}
+		}
+	}
+
+	if len(ngramsAscii) > 0 {
+		ngramsAscii = ngramsAscii[:len(ngramsAscii)-1]
+	}
+	if len(ngramsUnicode) > 0 {
+		ngramsUnicode = ngramsUnicode[:len(ngramsUnicode)-1]
+	}
+
+	asc := makeAsciiNgramOffset(ngramsAscii, offsetsAscii)
+	uni := makeArrayNgramOffset(ngramsUnicode, offsetsUnicode)
+
+	return combinedNgramOffset{asc, uni}
+}
+
+// Get returns a simpleSection with sz=0 if no entry, or sz=ngramAsciiMaxSectionLength
+// if the length of the ngram is too large for this type and it should cascade to the next entry.
+func (a combinedNgramOffset) Get(gram ngram) simpleSection {
+	var sec simpleSection
+	if gram&ngramAsciiMask == gram {
+		sec = a.asc.Get(gram)
+	}
+	if sec.sz == 0 || sec.sz == ngramAsciiMaxSectionLength {
+		sec = a.uni.Get(gram)
+	}
+	return sec
+}
+
+func (a combinedNgramOffset) DumpMap() map[ngram]simpleSection {
+	m := a.asc.DumpMap()
+	for k, v := range a.uni.DumpMap() {
+		m[k] = v
+	}
+	return m
+}
+
+func (a combinedNgramOffset) SizeBytes() int {
+	return a.asc.SizeBytes() + a.uni.SizeBytes()
+}
+
+const ngramAsciiMask = 127 | 127<<21 | 127<<42
+
+// Ascii mapping packs 3*7b chars and 11 bits of lengths, with this as the set maximum.
+// We could save another ~3% of total RAM / 5% of combinedNgramOffset RAM by switching to
+// a 40b packing with 19-bit lengths, but the code would be significantly uglier so it doesn't
+// seem worth it.
+const ngramAsciiMaxSectionLength = (1 << 11) - 1
+
+func ngramAsciiToPacked(ng ngram) uint32 {
+	return uint32(ng&127) | uint32((ng>>(21-7))&(127<<7)) | uint32((ng>>(42-14))&(127<<14))
+}
+
+func ngramAsciiPackedToNgram(ng uint32) ngram {
+	return ngram(ng&127) | ngram(ng&(127<<7))<<(21-7) | ngram(ng&(127<<14))<<(42-14)
+}
+
+// asciiNgramOffset stores ascii trigrams packed together with short lengths,
+// using offsets for a chunk of entries to limit the number of lengths that must
+// be summed to compute a section's offset.
+type asciiNgramOffset struct {
+	entries      []uint32 // (chara << 25 | charb << 18 | charc << 11 | length)
+	chunkOffsets []uint32 // offset for entries[i*asciiNgramOffsetChunkLength]
+}
+
+// asciiNgramOffsetChunkLength specifies how many entries share one initial offset.
+// It must be a power of 2, and was chosen empirically by measuring RAM usage:
+// 8: 4132MB, 16: 4047MB, 32: 4006MB, 64: 3992MB, 128: 3990MB
+const asciiNgramOffsetChunkLength = 32
+
+func makeAsciiNgramOffset(ngrams []uint32, offsets []uint32) *asciiNgramOffset {
```

Comment:
this seems very complicated, is it really worth it? I feel a bit reluctant to include this. I can understand it is an improvement, but it feels a bit early to introduce? I was going to approve this PR, but I'd rather we put this change into another PR (ie this is growing too much). Then we can discuss the merits of this approach without delaying the merging of all the good stuff you have done so far.

Reply:
5.2GB->3.8GB RAM for 226 lines + 87 lines of tests. I'll split it and discuss it separately.

Reply:
ok that is a great improvement. I'll review closely once its split out.